cpu/stm32f1/periph/gpio.c

```diff
@@ -30,1015 +30,190 @@
-            port = GPIO_2_PORT;
-            pin = GPIO_2_PIN;
-            break;
-#endif
-#if GPIO_3_EN
-        case GPIO_3:
-            GPIO_3_CLKEN();
-            port = GPIO_3_PORT;
-            pin = GPIO_3_PIN;
-            break;
-#endif
-#if GPIO_4_EN
-        case GPIO_4:
-            GPIO_4_CLKEN();
-            port = GPIO_4_PORT;
-            pin = GPIO_4_PIN;
-            break;
-#endif
-#if GPIO_5_EN
-        case GPIO_5:
-            GPIO_5_CLKEN();
-            port = GPIO_5_PORT;
-            pin = GPIO_5_PIN;
-            break;
-#endif
-#if GPIO_6_EN
-        case GPIO_6:
-            GPIO_6_CLKEN();
-            port = GPIO_6_PORT;
-            pin = GPIO_6_PIN;
-            break;
-#endif
-#if GPIO_7_EN
-        case GPIO_7:
-            GPIO_7_CLKEN();
-            port = GPIO_7_PORT;
-            pin = GPIO_7_PIN;
-            break;
-#endif
-#if GPIO_8_EN
-        case GPIO_8:
-            GPIO_8_CLKEN();
-            port = GPIO_8_PORT;
-            pin = GPIO_8_PIN;
-            break;
-#endif
-#if GPIO_9_EN
-        case GPIO_9:
-            GPIO_9_CLKEN();
-            port = GPIO_9_PORT;
-            pin = GPIO_9_PIN;
-            break;
-#endif
-#if GPIO_10_EN
-        case GPIO_10:
-            GPIO_10_CLKEN();
-            port = GPIO_10_PORT;
-            pin = GPIO_10_PIN;
-            break;
-#endif
-#if GPIO_11_EN
-        case GPIO_11:
-            GPIO_11_CLKEN();
-            port = GPIO_11_PORT;
-            pin = GPIO_11_PIN;
-            break;
-#endif
-#if GPIO_12_EN
-        case GPIO_12:
-            GPIO_12_CLKEN();
-            port = GPIO_12_PORT;
-            pin = GPIO_12_PIN;
-            break;
-#endif
-#if GPIO_13_EN
-        case GPIO_13:
-            GPIO_13_CLKEN();
-            port = GPIO_13_PORT;
-            pin = GPIO_13_PIN;
-            break;
-#endif
-#if GPIO_14_EN
-        case GPIO_14:
-            GPIO_14_CLKEN();
-            port = GPIO_14_PORT;
-            pin = GPIO_14_PIN;
-            break;
-#endif
-#if GPIO_15_EN
-        case GPIO_15:
-            GPIO_15_CLKEN();
-            port = GPIO_15_PORT;
-            pin = GPIO_15_PIN;
-            break;
-#endif
+/**
+ * @brief   Extract the port number from the given pin identifier
+ *
+ * Isolating bits 10 to 13 of the port base addresses leads to unique port
+ * numbers.
+ */
+static inline int _port_num(gpio_t pin)
+{
+    return ((pin >> 10) & 0x0f);
+}
 
-        default:
-            return -1;
-    }
+/**
+ * @brief   Get the pin number from the pin identifier, encoded in the LSB 4 bit
+ */
+static inline int _pin_num(gpio_t pin)
+{
+    return (pin & 0x0f);
+}
+
+
+int gpio_init(gpio_t pin, gpio_dir_t dir, gpio_pp_t pullup)
+{
+    GPIO_TypeDef *port = _port(pin);
+    int pin_num = _pin_num(pin);
 
+
+    /* enable the clock for the selected port */
+    RCC->APB2ENR |= (RCC_APB2ENR_IOPAEN << _port_num(pin));
+    /* clear configuration */
+    port->CR[pin_num >> 3] &= ~(0xf << ((pin_num & 0x7) * 4));
```

Comment:
Probably I'm just at loss currently. I'm wondering why you shift by `3` and not by `4` here. You want to distinguish between the `CR-low` and the `CR-high` registers (by indexing with `0` or `1`). And AFAIK you want to have `CR[0]` if `pin_num < 16` and  `CR[1]` if `pin_num >=16`. Now just two examples:

```
pin_num = 0d15 = 0b1111
pin_num >> 3 = 0d1 = 0b0001
```

```
pin_num = 0d16 = 0b10000
pin_num >> 3 = 0d2 = 0b0010
```

Where am I mistaken?